src/github_runner_image_builder/builder.py

```diff
@@ -0,0 +1,559 @@
+        UnsupportedArchitectureError: If an unsupported architecture was passed.
+
+    Returns:
+        The supported architecture.
+    """
+    match arch:
+        case Arch.X64:
+            return "amd64"
+        case Arch.ARM64:
+            return "arm64"
+        case _:
+            raise UnsupportedArchitectureError(arch)
+
+
+def _clean_build_state() -> None:
+    """Remove any artefacts left by previous build.
+
+    Raises:
+        CleanBuildStateError: if there was an error cleaning up the build state.
+    """
+    # The commands will fail if artefacts do not exist and hence there is no need to check the
+    # output of subprocess runs.
+    IMAGE_MOUNT_DIR.mkdir(parents=True, exist_ok=True)
+    try:
+        subprocess.run(
+            ["/usr/bin/umount", str(IMAGE_MOUNT_DIR / "dev")], timeout=30, check=False
+        )  # nosec: B603
+        subprocess.run(
+            ["/usr/bin/umount", str(IMAGE_MOUNT_DIR / "proc")], timeout=30, check=False
+        )  # nosec: B603
+        subprocess.run(
+            ["/usr/bin/umount", str(IMAGE_MOUNT_DIR / "sys")], timeout=30, check=False
+        )  # nosec: B603
+        subprocess.run(
+            ["/usr/bin/umount", str(IMAGE_MOUNT_DIR)], timeout=30, check=False
+        )  # nosec: B603
+        subprocess.run(
+            ["/usr/bin/umount", str(NETWORK_BLOCK_DEVICE_PATH)], timeout=30, check=False
+        )  # nosec: B603
+        subprocess.run(  # nosec: B603
+            ["/usr/bin/umount", str(NETWORK_BLOCK_DEVICE_PARTITION_PATH)], timeout=30, check=False
+        )
+        subprocess.run(  # nosec: B603
+            ["/usr/bin/qemu-nbd", "--disconnect", str(NETWORK_BLOCK_DEVICE_PATH)],
+            timeout=30,
+            check=False,
+        )
+        subprocess.run(  # nosec: B603
+            ["/usr/bin/qemu-nbd", "--disconnect", str(NETWORK_BLOCK_DEVICE_PARTITION_PATH)],
+            timeout=30,
+            check=False,
+        )
+    except subprocess.SubprocessError as exc:
+        raise CleanBuildStateError from exc
+
+
+def _download_cloud_image(arch: Arch, base_image: BaseImage) -> Path:
+    """Download the cloud image from cloud-images.ubuntu.com.
+
+    Args:
+        arch: The cloud image architecture to download.
+        base_image: The ubuntu base image OS to download.
+
+    Returns:
+        The downloaded cloud image path.
+
+    Raises:
+        CloudImageDownloadError: If there was an error downloading the image.
+    """
+    try:
+        bin_arch = _get_supported_runner_arch(arch)
+    except UnsupportedArchitectureError as exc:
+        raise CloudImageDownloadError from exc
+
+    try:
+        # The ubuntu-cloud-images is a trusted source
+        image_path, _ = urllib.request.urlretrieve(  # nosec: B310
+            CLOUD_IMAGE_URL_TMPL.format(BASE_IMAGE=base_image.value, BIN_ARCH=bin_arch),
+            CLOUD_IMAGE_FILE_NAME_TMPL.format(BASE_IMAGE=base_image.value, BIN_ARCH=bin_arch),
+        )
+        return Path(image_path)
```

Comment:
perhaps make the image_path variable first

Reply:
updated, thank you!